### PR TITLE
Adversarial liquidation added

### PR DESCRIPTION
**Pull request by msabvid**
- Adversarial liquidator added. 
- Batch runner for both examples
- Postprocessing of results of batch runner

### CODE REVIEW

File: simulations/agents/uniswap_agent.py [hunk 1]
```diff
@@ -183,7 +182,7 @@ def _quote_price(change_token_1):
                     maxiter=5,
                 )
                 change_token_1 = sol.root
-            except eth_abi.exceptions.ValueOutOfBounds:
+            except:  # noqa: E722
```
Comment (zombie-einstein) [inline simulations/agents/uniswap_agent.py:185]: Just so I know, how come this is reverted?

File: simulations/uniswap/postprocessing.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import os
+from functools import reduce
+from typing import List, Tuple
+
+import pandas as pd
+
+
+def save(batch_results: List[List[Tuple[int, int]]]):
```
Comment (zombie-einstein) [inline simulations/uniswap/postprocessing.py:8]: Is this function the same as the aave version (apart from the path), could be consolidated into one?

File: uniswap_example.py [hunk 3]
```diff
@@ -10,8 +16,33 @@
...
+    )
     args = parser.parse_args()
 
+    # run a single simulation
     results = simulations.uniswap.sim.run_from_cache(args.seed, args.n_steps)
-
     simulations.uniswap.plotting.plot_results(results)
+
+    # run a batch of simulations
+    if args.batch_runner:
```
Comment (zombie-einstein) [inline uniswap_example.py:31]: Also does this need an `else`?

File: uniswap_example.py [hunk 4]
```diff
@@ -10,8 +16,33 @@
...
+        parameters_samples = [
+            dict(mu=mu, sigma=sigma)
+            for mu, sigma in product([0.0, 0.1, -0.1], [0.1, 0.2, 0.3])
+        ]
+
+        with open(os.path.join("simulations", "uniswap", "cache.json"), "r") as f:
+            cache_json = json.load(f)
+        cache = verbs.utils.cache_from_json(cache_json)
```
Comment (zombie-einstein) [inline uniswap_example.py:39]: Do we not have a cache from file function? If not we should add

File: simulations/uniswap/sim.py [hunk 5]
```diff
@@ -110,7 +117,12 @@ def runner(env, seed: int, n_steps: int, init_cache: bool = False):
     runner = verbs.sim.Sim(seed, env, agents)
     results = runner.run(n_steps=n_steps)
 
-    return env, results
+    if init_cache:
```
[the remark this answers is not in the document]
Comment (msabvid) [reply simulations/uniswap/sim.py:120]: If the runner returns `env, results`, then the joblib library in `batch_runner` tries to pickle `env` and it crashes. So I'm just returning `results` right now, and saving the environment cache in the `runner` if it is the initial simulation

File: simulations/agents/uniswap_agent.py (hunk 1 above)
Comment (msabvid) [reply simulations/agents/uniswap_agent.py:185]: One of the revert error was that the trade quoted was negative and the eth_abi could not encode the data. That was the eth_abi.exceptions.ValueOutOfBounds. 

Another revert error was from the `call()` and it was an empty error. I don't have full control over the iterations of the  `root_scalar` algorithm. In this case, when it crashes I just return an empty transaction

File: simulations/uniswap/postprocessing.py (hunk 2 above)
Comment (msabvid) [reply simulations/uniswap/postprocessing.py:8]: Done in a new commit

File: uniswap_example.py (hunk 3 above)
Comment (msabvid) [reply uniswap_example.py:31]: Solved in a new commit

File: uniswap_example.py [hunk 6]
```diff
@@ -10,8 +16,33 @@
...
+
+        batch_results = batch_run(
+            simulations.uniswap.sim.runner,
+            n_steps=100,
```
[the remark this answers is not in the document]
Comment (msabvid) [reply uniswap_example.py:43]: Solved in a new commit

File: uniswap_example.py (hunk 4 above)
Comment (msabvid) [reply uniswap_example.py:39]: I don't think so. Will add in new branch of verbs

File: simulations/uniswap/sim.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply simulations/uniswap/sim.py:120]: Ah see what you are saying. Is it a bit neater to have a runner that returns both, and then specialize it for the cache-creation and execution use cases?

File: aave_example.py [hunk 7]
```diff
@@ -26,3 +37,23 @@
     )
 
     simulations.aave.plotting.plot_results(results, args.n_borrow_agents)
+
+    # run a batch of simulations
+    if args.batch_runner:
```
[the remark this answers is not in the document]
Comment (msabvid) [reply aave_example.py:35]: solved in a new commit

File: simulations/uniswap/sim.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (msabvid) [reply simulations/uniswap/sim.py]: fixed in new commit

**Pull request merged**